validator/src/machine/consensus/rollover.test.ts

```diff
@@ -0,0 +1,248 @@
+	x: 73844941487532555987364396775795076447946974313865618280135872376303125438365n,
+	y: 29462187596282402403443212507099371496473451788807502182979305411073244917417n,
+});
+
+const MACHINE_CONFIG: MachineConfig = {
+	defaultParticipants: [
+		{
+			id: 1n,
+			address: zeroAddress,
+		},
+		{
+			id: 3n,
+			address: zeroAddress,
+		},
+		{
+			id: 7n,
+			address: zeroAddress,
+		},
+	],
+	genesisSalt: zeroHash,
+	keyGenTimeout: 20n,
+	signingTimeout: 0n,
+	blocksPerEpoch: 10n,
+};
+
+const CONSENSUS_STATE: ConsensusState = {
+	activeEpoch: 0n,
+	stagedEpoch: 0n,
+	genesisGroupId: "0x5af35afe",
+	groupPendingNonces: {},
+	epochGroups: {},
+	signatureIdToMessage: {},
+};
+
+// By default we setup in a genesis state
+// This avoids that nonce commitments are triggered every time
+const MACHINE_STATES: MachineStates = {
+	rollover: {
+		id: "waiting_for_rollover",
+	},
+	signing: {},
+};
+
+// --- Tests ---
+describe("check rollover", () => {
+	it("should not trigger key gen in genesis state", async () => {
+		const protocol = {} as unknown as ShieldnetProtocol;
+		const keyGenClient = {} as unknown as KeyGenClient;
+		const consensus: ConsensusState = {
+			...CONSENSUS_STATE,
+			genesisGroupId: undefined,
+		};
+		const diff = checkEpochRollover(MACHINE_CONFIG, protocol, keyGenClient, consensus, MACHINE_STATES, 1n);
+
+		expect(diff).toStrictEqual({});
+	});
+
+	it("should not trigger key gen if not waiting for rollover", async () => {
+		const protocol = {} as unknown as ShieldnetProtocol;
+		const keyGenClient = {} as unknown as KeyGenClient;
+		const machineStates: MachineStates = {
+			...MACHINE_STATES,
+			rollover: {
+				id: "collecting_commitments",
+				groupId: "0xda5afe3",
+				nextEpoch: 2n,
+				deadline: 22n,
+			},
+		};
+		const diff = checkEpochRollover(MACHINE_CONFIG, protocol, keyGenClient, CONSENSUS_STATE, machineStates, 1n);
+
+		expect(diff).toStrictEqual({});
+	});
+
+	it("should not trigger key gen if rollover is staged", async () => {
+		const protocol = {} as unknown as ShieldnetProtocol;
+		const keyGenClient = {} as unknown as KeyGenClient;
+		const consensus: ConsensusState = {
+			...CONSENSUS_STATE,
+			stagedEpoch: 1n,
+		};
+		const diff = checkEpochRollover(MACHINE_CONFIG, protocol, keyGenClient, consensus, MACHINE_STATES, 1n);
+
+		expect(diff).toStrictEqual({});
+	});
+
+	it("should update staged rollover without triggering key gen", async () => {
+		const protocol = {} as unknown as ShieldnetProtocol;
+		const keyGenClient = {} as unknown as KeyGenClient;
+		const machineStates: MachineStates = {
+			...MACHINE_STATES,
+			rollover: {
+				id: "collecting_commitments",
+				groupId: "0xda5afe3",
+				nextEpoch: 2n,
+				deadline: 22n,
+			},
+		};
+		const consensus: ConsensusState = {
+			...CONSENSUS_STATE,
+			stagedEpoch: 1n,
+		};
+		const diff = checkEpochRollover(MACHINE_CONFIG, protocol, keyGenClient, consensus, machineStates, 10n);
+
+		expect(diff.actions).toBeUndefined();
```

Comment:
TBH, here, I would expect a `KeyGen` action right away (at least, this is my understanding of how this is implemented in the spec). At least, this is how I understood these tests.